lib/worldwide/address.rb

```diff
@@ -463,5 +475,32 @@ def concatenation_delimiter(concatenated_field, decorator, field_index)
         concatenated_field.blank? ? RESERVED_DELIMITER : "#{decorator}#{RESERVED_DELIMITER}"
       end
     end
+
+    def parse_fields(field_definitions, field_values)
+      field_definitions.each_with_index.with_object({}) do |(field, i), parsed_fields|
+        field_name = field["key"].underscore
+        field_value = if i < field_values.length
+          raw_field_value = field_values[i]
+          next_field = field_definitions[i + 1] unless last_field_in_arr?(field_definitions, i)
+          strip_decorators(next_field, raw_field_value)
+        else
+          ""
+        end
+        parsed_fields[field_name] = field_value
+      end
+    end
+
+    def last_field_in_arr?(arr, i)
+      i == arr.length - 1
+    end
+
+    def strip_decorators(field_definition, raw_field_value)
```

Comment:
I recommend we rename the first param to `next_field_definition` and the second one to `current_field_value`, to clarify their relation.

Also, I think `String#delete_suffix` should work here.

```ruby
def strip_decorators(next_field_definition, current_field_value)
  delimiter = next_field_definition&.fetch("decorator", nil) || ""
  current_field_value.delete_suffix(delimiter)
end
```